xbmc/cores/RetroPlayer/playback/GameLoop.cpp

```diff
@@ -56,75 +56,58 @@ void CGameLoop::Process(void)
 {
   while (!m_bStop)
   {
-    if (m_speedFactor == 0.0)
+    if (m_loopSpeedFactor != m_speedFactor)
     {
-      m_lastFrameMs = 0.0;
+      m_lastFrameUs = std::chrono::microseconds::zero();
+      m_loopSpeedFactor = m_speedFactor;
+    }
+
+    if (m_loopSpeedFactor == 0.0)
+    {
+      m_lastFrameUs = std::chrono::microseconds::zero();
       m_sleepEvent.Wait(5000ms);
     }
     else
     {
-      if (m_speedFactor > 0.0)
+      if (m_lastFrameUs == std::chrono::microseconds::zero())
+        m_lastFrameUs = NowUs();
+
+      if (m_loopSpeedFactor > 0.0)
         m_callback->FrameEvent();
-      else if (m_speedFactor < 0.0)
+      else if (m_loopSpeedFactor < 0.0)
         m_callback->RewindEvent();
 
-      if (m_lastFrameMs > 0.0)
-      {
-        m_lastFrameMs += FrameTimeMs();
-        m_adjustTime = m_lastFrameMs - NowMs();
-      }
-      else
-      {
-        m_lastFrameMs = NowMs();
-        m_adjustTime = 0.0;
-      }
-
       // Calculate sleep time
-      double sleepTimeMs = SleepTimeMs();
+      std::chrono::microseconds sleepTimeUs = SleepTimeUs();
 
-      // Sleep at least 1 ms to avoid sleeping forever
-      while (sleepTimeMs > 1.0)
-      {
-        m_sleepEvent.Wait(std::chrono::milliseconds(static_cast<unsigned int>(sleepTimeMs)));
-
-        if (m_bStop)
-          break;
-
-        // Speed may have changed, update sleep time
-        sleepTimeMs = SleepTimeMs();
-      }
+      if (sleepTimeUs > std::chrono::microseconds::zero())
+        m_sleepEvent.Wait(sleepTimeUs);
     }
   }
 }
 
-double CGameLoop::FrameTimeMs() const
+std::chrono::microseconds CGameLoop::FrameTimeUs() const
 {
-  if (m_speedFactor != 0.0)
-    return 1000.0 / m_fps / std::abs(m_speedFactor);
+  if (m_loopSpeedFactor != 0.0)
+    return std::chrono::duration_cast<std::chrono::microseconds>(1s / m_fps /
+                                                                 std::abs(m_loopSpeedFactor));
   else
-    return 1000.0 / m_fps / 1.0;
+    return std::chrono::duration_cast<std::chrono::microseconds>(1s / m_fps);
 }
 
-double CGameLoop::SleepTimeMs() const
+std::chrono::microseconds inline CGameLoop::SleepTimeUs()
 {
   // Calculate next frame time
-  const double nextFrameMs = m_lastFrameMs + FrameTimeMs();
+  const std::chrono::microseconds nextFrameUs = m_lastFrameUs += FrameTimeUs();
```

Comment:
Lack of parenthesis here makes it harder to follow order of operations, and prefix vs postfix incrementation. Can it be made more clear?

Reply:
Applied and removed a superfluous line.

Reply:
Much better. I would take it a step further and separate the compound statement into two lines. Such as:

```patch
--- a/xbmc/cores/RetroPlayer/playback/GameLoop.cpp
+++ b/xbmc/cores/RetroPlayer/playback/GameLoop.cpp
@@ -75,7 +75,13 @@ void CGameLoop::Process(void)
       else if (m_loopSpeedFactor < 0.0)
         m_callback->RewindEvent();
 
-      std::chrono::microseconds nextFrameUs = (m_lastFrameUs += FrameTimeUs());
+      // Calculate the time for the next frame by adding the frame duration to
+      // the last frame time
+      const std::chrono::microseconds nextFrameUs = m_lastFrameUs + frameDurationUs;
+
+      // Update the last frame time to ensure accurate timing for the next loop
+      // iteration
+      m_lastFrameUs = nextFrameUs;
 
       std::chrono::microseconds sleepTimeUs = (nextFrameUs - NowUs());
```

Reply:
I've added this change to my patch below